server/src/main/java/org/elasticsearch/script/JodaCompatibleZonedDateTime.java

```diff
@@ -0,0 +1,379 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.script;
+
+import org.elasticsearch.common.logging.DeprecationLogger;
+import org.elasticsearch.common.logging.ESLoggerFactory;
+import org.joda.time.DateTime;
+import org.joda.time.DateTimeZone;
+
+import java.security.AccessController;
+import java.security.PrivilegedAction;
+import java.time.DayOfWeek;
+import java.time.Instant;
+import java.time.LocalDate;
+import java.time.LocalDateTime;
+import java.time.Month;
+import java.time.OffsetDateTime;
+import java.time.ZoneId;
+import java.time.ZonedDateTime;
+import java.time.temporal.ChronoField;
+import java.time.temporal.TemporalAdjuster;
+import java.time.temporal.TemporalAmount;
+import java.time.temporal.TemporalField;
+import java.time.temporal.TemporalUnit;
+import java.time.temporal.WeekFields;
+import java.util.Locale;
+
+/**
+ * A wrapper around ZonedDateTime that exposes joda methods for backcompat.
+ */
+public class JodaCompatibleZonedDateTime {
+    private static final DeprecationLogger DEPRECATION_LOGGER =
+        new DeprecationLogger(ESLoggerFactory.getLogger(JodaCompatibleZonedDateTime.class));
+
+    private static void logDeprecated(String key, String message, Object... params) {
+        AccessController.doPrivileged((PrivilegedAction<Void>) () -> {
```

Comment:
It is worth a comment about why this doesn't check special permission: namely that we expect this to be called by scripts which won't have it.